src/ssl.c

```diff
@@ -51306,6 +51303,597 @@ int wolfSSL_PEM_write_bio_PKCS7(WOLFSSL_BIO* bio, PKCS7* p7)
     }
     return WOLFSSL_FAILURE;
 }
+
+/*****************************************************************************
+* wolfSSL_SMIME_read_PKCS7 - Reads the given S/MIME message and parses it into
+* a PKCS7 object. In case of a multipart message, stores the signed data in
+* bcont.
+*
+* RETURNS:
+* returns pointer to a PKCS7 structure on success, otherwise returns NULL
+*/
+WOLFSSL_API PKCS7* wolfSSL_SMIME_read_PKCS7(WOLFSSL_BIO* in,
+                                            WOLFSSL_BIO** bcont)
+{
+    MimeHdr* allHdrs = NULL;
+    MimeHdr* curHdr = NULL;
+    MimeParam* curParam = NULL;
+    int inLen = 0;
+    int sectionLen = 0;
+    int ret = -1;
+    char* section = NULL;
+    PKCS7* pkcs7 = NULL;
+    word32 outLen = 0;
+    byte* out = NULL;
+    byte* outHead = NULL;
+
+
+    if (!in) {
+        return NULL;
+    }
+    inLen = wolfSSL_BIO_get_len(in);
+    if (inLen <= 0) {
+        return NULL;
+    }
+
+    ret = MIME_parse_headers(in, inLen, &allHdrs);
+    if (ret < 0) {
+        WOLFSSL_MSG("Parsing MIME headers failed.\n");
+        MIME_free_hdrs(allHdrs);
+        return NULL;
+    }
+
+    curHdr = MIME_find_header_name("Content-Type", allHdrs);
+    if (curHdr && !XSTRNCMP(curHdr->body, "multipart/signed",
+                            XSTRLEN("multipart/signed"))) {
```

Comment:
Since these are static strings then using `XSTR_SIZEOF` would be more appropriate. This would remove the run-time string length counting.

Reply:
Good idea, thanks, updated.